[FEATURE] Implement bar chart panel plugin

ellenwyllie opened the pull request:
<img width="651" alt="Screen Shot 2023-07-18 at 10 36 50 AM" src="https://github.com/perses/perses/assets/22085207/3ac8e493-4024-4eaf-9314-beb3a8e0e499">
<img width="651" alt="Screen Shot 2023-07-18 at 10 47 08 AM" src="https://github.com/perses/perses/assets/22085207/14fd7ba9-b218-4c84-b43e-122cc5d2d7ae">

### Sorting
The data can be sorted in ascending or descending order. By default the bars are sorted in descending order.
<img width="1077" alt="Screen Shot 2023-07-18 at 11 22 34 AM" src="https://github.com/perses/perses/assets/22085207/3e756883-7a21-446b-90fe-f946828479fa">


### Percentage Mode
When Percentage is selected from the Mode dropdown, the values are calculated as a percentage. Note that the unit selector is disabled when in Percentage mode.
<img width="1075" alt="Screen Shot 2023-07-19 at 10 37 27 AM" src="https://github.com/perses/perses/assets/22085207/5acd7958-2de1-4821-b31f-d55aca89a893">


Percentage mode is disabled when the unit is a percent:
<img width="1074" alt="Screen Shot 2023-07-18 at 11 03 41 AM" src="https://github.com/perses/perses/assets/22085207/341eb76d-6629-412f-9fc7-dedbb20a4ebc">



# Checklist

- [ ] Pull request has a descriptive title and context useful to a reviewer.
- [ ] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [ ] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [ ] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [ ] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [ ] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

eunicorn commented:
When there are many series and panel size isn't big enough, it becomes hard to read since the bars and numbers are crowded and overlap each other. Maybe there should there a min height of the bars and it starts scrolling?

![image](https://github.com/perses/perses/assets/9817826/15028b9e-6b1c-4ec1-9bd3-bd5f77b98434)

eunicorn commented on ui/components/src/BarChart/BarChart.tsx:
```diff
@@ -32,14 +33,15 @@ export interface BarChartProps {
   height: number;
   data: BarChartData[] | null;
   unit: UnitOptions;
+  mode: ModeOption;
```
should this be undefined and make `Value` the default?

eunicorn commented on ui/panels-plugin/src/plugins/bar-chart/bar-chart-model.ts:
```diff
@@ -0,0 +1,49 @@
+export interface BarChartOptions {
+  calculation: CalculationType;
+  unit: UnitOptions;
+  sort: SortOption;
```
I think sort can also be optional and we can make 'asc' as default

eunicorn replied:
In general, i think we want as many defaults as possible so we don't have to explicitly define every option in the json schema

ellenwyllie replied:
Makes sense, I updated `unit`, `sort` and `mode` to be optional! ef0a2538ce1f30459e24f92dfd1f01ef2c9207de

eunicorn replied:
I'm getting some build errors, I think you also need to update the props in ModeSelector and SortSelector

ellenwyllie replied:
Oops, should be good now

sjcobb replied:
Not necessary for this PR, but I like the 'Reset to Defaults' button for this too to make it easy to revert to default spec, it should really be handled by the framework in the future and on StatChart and GaugeChart too
<img width="306" alt="image" src="https://github.com/perses/perses/assets/9369625/1782a1ee-9688-4174-bd5b-9e6c320fcf1d">

eunicorn commented on ui/panels-plugin/src/plugins/bar-chart/utils.test.ts:
```diff
@@ -0,0 +1,104 @@
+  },
+];
+
+describe('calculatePercentages', () => {
```
Yay thanks for adding tests. Maybe also good to add test cases with null values?

ellenwyllie replied:
Good idea, updated the tests dfd71b6603b5cca95f0bbe8dbcfa137375d4f47d

eunicorn commented on ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:
```diff
@@ -0,0 +1,89 @@
+      return [];
+    }
+    const seriesData: BarChartData[] = [];
+    for (const timeSeries of queryResults[0].data.series) {
```
Actually i think we need to loop through `queryResults` here since you can have multiple queries, and each query can also have multiple series

eunicorn replied:
You can find an example using the LegendExample panel in Benchmark dashboard

ellenwyllie replied:
Updated! f9dc6a03c675261b3259dba007af55aabe3ec478

ellenwyllie commented:
> When there are many series and panel size isn't big enough, it becomes hard to read since the bars and numbers are crowded and overlap each other. Maybe there should there a min height of the bars and it starts scrolling?

Good point! I just added a min height for the bars: 8b2b30fafca5c028d3438a84ed8dc89eb839b594

https://github.com/perses/perses/assets/22085207/b133b596-758f-404f-b36d-3c279d013a27

ellenwyllie commented on ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  return (
+    <Stack
```
Removed the `Stack` here 0f1879ea0779d506fd1af602689c184b5cca9203

ellenwyllie commented on ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:
```diff
@@ -0,0 +1,89 @@
+import { BarChartOptions } from './bar-chart-model';
+import { calculatePercentages, sortSeriesData } from './utils';
+
+const PADDING = 24;
```
Updated! 0f1879ea0779d506fd1af602689c184b5cca9203

eunicorn commented:
@ellenwyllie one last thing, there is a merge conflict that you need to resolve

sjcobb commented on ui/panels-plugin/src/plugins/bar-chart/BarChartPanel.tsx:
```diff
@@ -0,0 +1,86 @@
+
+  return (
+    <Box sx={{ padding: `${PADDING}px` }}>
+      <BarChart
```
We might need some different options for empty state like in TimeSeriesChart, when adding a brand new BarChart, it looks a little odd how the axisLine shows on the left like this: 

<img width="986" alt="image" src="https://github.com/perses/perses/assets/9369625/9321f446-814f-4f48-9de3-aec009233bfd">

At first I wasn't sure what that line was, took a second to remember we default to horizontal orientation with `yAxis.type: 'category'`

sjcobb commented on schemas/panels/bar/bar.cue:
```diff
@@ -0,0 +1,26 @@
+	calculation: common.#calculation
+	unit?:       common.#unit
+	sort?:       "asc" | "desc"
+	mode?:       "value" | "percentage"
```
Do we know roughly how this schema maps to Grafana's [bargauge](https://play.grafana.org/d/ktMs4D6Mk/5-bar-charts-and-pie-charts?orgId=1&editPanel=13&showCategory=Value%20options) supported options?

<img width="760" alt="image" src="https://github.com/perses/perses/assets/9369625/8b4ccda3-2f55-4292-8d79-48d411f2ea23">

ellenwyllie replied:
I don't see a percent mode or anything similar for Grafana's bar gauge. Maybe when we do the conversion we just have to switch it back to the raw values?

sjcobb commented on ui/components/src/BarChart/BarChart.tsx:
```diff
@@ -31,11 +36,12 @@ export interface BarChartProps {
+  const { width, height, data, unit = { kind: 'Decimal' }, mode = 'value' } = props;
   const chartsTheme = useChartsTheme();
 
   const option: EChartsCoreOption = useMemo(() => {
```
Seeing some cutoff with longer axis labels, adjusting grid.left or one of the axisLabel options might help

<img width="596" alt="image" src="https://github.com/perses/perses/assets/9369625/455efcea-f961-41a1-ade0-4e01f246c3b2">

sjcobb replied:
I think we should adjust default to show 'M', 'B' notation on value axis like TimeSeriesChart 

<img width="630" alt="image" src="https://github.com/perses/perses/assets/9369625/e31b194f-2ec9-4ca6-8772-de9ef54d3668">

sjcobb replied:
Noticed some cutoff on right side when testing percent mode, I think those [grid](https://echarts.apache.org/en/option.html#grid) properties will help, surprised the theme defaults don't cover us though 😕 

<img width="962" alt="image" src="https://github.com/perses/perses/assets/9369625/cda57bf0-4075-4333-9668-99cb4a2aedf6">

sjcobb reviewed (approved):
LGTM! I left some smaller comments but they could just be done in a follow up and I'll get this into v0.38.0 and do a patch release later with the rest

ellenwyllie commented:
> LGTM! I left some smaller comments but they could just be done in a follow up and I'll get this into v0.38.0 and do a patch release later with the rest

@sjcobb I addressed your comments in this PR: https://github.com/perses/perses/pull/1325
